File: src/android/PolarBlePlugin.kt
package polarbleplugin

import android.util.Log
import android.widget.Toast
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.PolarBleApiDefaultImpl
import com.polar.sdk.api.PolarBleApiCallback
import com.polar.sdk.api.model.PolarDeviceInfo
import com.polar.sdk.api.model.PolarEcgData
import com.polar.sdk.api.model.PolarHrData
import com.polar.sdk.api.model.PolarSensorSetting
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.Disposable
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaInterface
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.CordovaWebView
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID


class PolarBlePlugin : CordovaPlugin() {

    private val variablesMap = mutableMapOf<String, String>()
    private val callbackMap = mutableMapOf<String, CallbackContext>()

    private lateinit var api: PolarBleApi
    private lateinit var deviceId: String
    private var ecgDisposable: Disposable? = null
    private var hrDisposable: Disposable? = null

    override fun initialize(cordova: CordovaInterface, webView: CordovaWebView) {
        super.initialize(cordova, webView)
    }

    override fun execute(action: String, args: JSONArray, callbackContext: CallbackContext): Boolean {
        try {
            when (action) {
                "connect" -> {
                    Log.d(TAG, "connect to device")
                    deviceId = args.getString(0)
                    api = PolarBleApiDefaultImpl.defaultImplementation(
                        cordova.context.applicationContext ,
                        setOf(
                            PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ONLINE_STREAMING,
                            PolarBleApi.PolarBleSdkFeature.FEATURE_BATTERY_INFO,
                            PolarBleApi.PolarBleSdkFeature.FEATURE_DEVICE_INFO
                        )
                    )

                    api.setApiCallback(object : PolarBleApiCallback() {
                        override fun blePowerStateChanged(powered: Boolean) {
                            Log.d(TAG, "BLE power: $powered")
                        }

                        override fun deviceConnected(polarDeviceInfo: PolarDeviceInfo) {
                            Log.d(TAG, "CONNECTED: ${polarDeviceInfo.deviceId}")
                        }

                        override fun deviceConnecting(polarDeviceInfo: PolarDeviceInfo) {
                            Log.d(TAG, "CONNECTING: ${polarDeviceInfo.deviceId}")
                        }

                        override fun deviceDisconnected(polarDeviceInfo: PolarDeviceInfo) {
                            Log.d(TAG, "DISCONNECTED: ${polarDeviceInfo.deviceId}")
                        }

                        override fun bleSdkFeatureReady(
                            identifier: String,
                            feature: PolarBleApi.PolarBleSdkFeature
                        ) {
                            Log.d(TAG, "Polar BLE SDK feature $feature is ready")
                            when (feature) {
                                PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_ONLINE_STREAMING -> {
                                    streamECG(callbackContext)
                                    streamHR(callbackContext)
                                }
                                else -> {}
                            }
                        }

                        override fun disInformationReceived(identifier: String, uuid: UUID, value: String) {
                            Log.d(TAG, "DIS INFO uuid: $uuid value: $value")
                        }

                        override fun batteryLevelReceived(identifier: String, level: Int) {
                            Log.d(TAG, "BATTERY LEVEL: $level")
                            variablesMap["batteryLevel"] = level.toString()
                            changeNotification("batteryLevel")
                        }
                    })
                    api.connectToDevice(args.getString(0))
                    callbackContext.success("connected")
                }
                "disconnect" -> {
                    api.disconnectFromDevice(args.getString(0))
                    callbackContext.success("disconnected");
                }
                "setCallback" -> {
                    val variableId = args.optString(0, "")
                    val callbackId = args.optString(1, "")
                    callbackMap["$variableId:$callbackId"] = callbackContext
                    return true
                }
                "echo" -> {
                    val input = args.getString(0)
                    val output = "Kotlin says \"$input\""
                    Toast.makeText(webView.context, output, Toast.LENGTH_LONG).show()
                    callbackContext.success(output)
                }
                else -> handleError("Invalid action", callbackContext)
            }
            return true
        } catch (e: Exception) {
            handleException(e, callbackContext)
            return false
        }
    }

    private fun sendValue(variableId: String, callbackContext: CallbackContext) {
        val valor = variablesMap[variableId] ?: ""
        val result = PluginResult(PluginResult.Status.OK, valor)
        result.keepCallback = true
        callbackContext.sendPluginResult(result)
    }

    private fun changeNotification(variableId: String) {
        val valor = variablesMap[variableId] ?: ""
        val res = PluginResult(PluginResult.Status.OK, valor)
        res.keepCallback = true

        callbackMap.keys.filter { it.startsWith("$variableId:") }.forEach {
            callbackMap[it]?.sendPluginResult(res)
        }
    }

    private fun streamHR(callbackContext: CallbackContext) {
        val isDisposed = hrDisposable?.isDisposed ?: true
        if (isDisposed) {
            hrDisposable = api.startHrStreaming(deviceId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { hrData: PolarHrData ->
                        for (sample in hrData.samples) {
                            //Log.d(TAG, "HR " + sample.hr)
                            //if (sample.rrsMs.isNotEmpty()) {
                                //val rrText = "(${sample.rrsMs.joinToString(separator = "ms, ")}ms)"
                                //textViewRR.text = rrText
                            //}
                            variablesMap["heartRate"] = sample.hr.toString()
                            changeNotification("heartRate")
                        }
                    },
                    { error: Throwable ->
                        Log.e(TAG, "HR stream failed. Reason $error")
                        hrDisposable = null
                    },
                    { Log.d(TAG, "HR stream complete") }
                )
        } else {
            // NOTE stops streaming if it is "running"
            hrDisposable?.dispose()
            hrDisposable = null
        }
    }

    private fun streamECG(callbackContext: CallbackContext) {
        val isDisposed = ecgDisposable?.isDisposed ?: true
        if (isDisposed) {
            ecgDisposable = api.requestStreamSettings(deviceId, PolarBleApi.PolarDeviceDataType.ECG)
                .toFlowable()
                .flatMap { sensorSetting: PolarSensorSetting -> api.startEcgStreaming(deviceId, sensorSetting.maxSettings()) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { polarEcgData: PolarEcgData ->
                        handleECGData(polarEcgData, callbackContext)
                    },
                    { error: Throwable ->
                        Log.e(TAG, "Ecg stream failed $error")
                        ecgDisposable = null
                    },
                    {
                        Log.d(TAG, "Ecg stream complete")
                    }
                )
        } else {
            // NOTE stops streaming if it is "running"
            ecgDisposable?.dispose()
            ecgDisposable = null
        }
    }

    private fun handleECGData(polarEcgData: PolarEcgData, callbackContext: CallbackContext) {
        val timeData: List<Long> = polarEcgData.samples.map { it.timeStamp }
        val waveData: List<Float> = polarEcgData.samples.map { it.voltage.toFloat() / 1000.0.toFloat() }

        val ecg = JSONObject()
        ecg.put("time", JSONArray(timeData))
        ecg.put("data", JSONArray(waveData))
		
		//var delta = (timeData.last() - timeData.first()) / timeData.size // nanoseconds
        //var frequency = (1.0 / delta) * 1000 * 1000 * 1000
        //Log.d(TAG, "Frequency: " +  frequency.toString())

        variablesMap["ecg"] = ecg.toString()
        changeNotification("ecg")
    }

    private fun handleError(errorMsg: String, callbackContext: CallbackContext) {
        Log.e(TAG, errorMsg)
        callbackContext.error(errorMsg)
    }

    private fun handleException(exception: Exception, callbackContext: CallbackContext) {
        handleError(exception.toString(), callbackContext)
    }

    companion object {
        private const val TAG = "PolarBlePlugin"
    }

}

File: src/ios/PolarBlePlugin.swift
import PolarBleSdk
import RxSwift
import Foundation
import CoreBluetooth


@objc(PolarBlePlugin)
class PolarBlePlugin: CDVPlugin, ObservableObject {
    
    var api: PolarBleApi!
    var deviceId: String!
    var variablesMap = [String: String]()
    var callbackMap = [String: String]()
    private let disposeBag = DisposeBag()
    var hrDisposable: Disposable?
    var ecgDisposable: Disposable?

    @objc(connect:)
    func connect(command: CDVInvokedUrlCommand) {
        var pluginResult: CDVPluginResult
        
        guard let arg = command.arguments[0] as? String else {
            pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
            commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        deviceId = arg
        
        api = PolarBleApiDefaultImpl.polarImplementation(DispatchQueue.main,
                                                         features: [PolarBleSdkFeature.feature_battery_info,
                                                                    PolarBleSdkFeature.feature_device_info,
                                                                    PolarBleSdkFeature.feature_polar_online_streaming])
        
        
        api.deviceInfoObserver = self
        api.deviceFeaturesObserver = self
        
        var echoedMessage: String = ""
        do {
            try api.connectToDevice(deviceId)
            NSLog("POLAR API CONNECTED")
            echoedMessage = "POLAR API CONNECTED"
        } catch {
            NSLog("ERROR CONNECTING API")
            echoedMessage = "ERROR CONNECTING API"
        }
        
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: echoedMessage)
        commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(disconnect:)
    func disconnect(command: CDVInvokedUrlCommand) {
        var pluginResult: CDVPluginResult
        
        guard let deviceId = command.arguments[0] as? String else {
            pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
            commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        do {
            try api.disconnectFromDevice(deviceId)
        } catch let err {
            NSLog("Failed to disconnect from \(deviceId). Reason \(err)")
        }
        
        let message = "disconnected"
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(setCallback:)
    func setCallback(command: CDVInvokedUrlCommand) {
        var pluginResult: CDVPluginResult
        
        guard let variableId = command.arguments[0] as? String,
              let callbackId = command.arguments[1] as? String else {
            pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
            commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        let key = "\(variableId):\(callbackId)"
        callbackMap[key] = command.callbackId
        
        //pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "kk")
        //commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(echo:)
    func echo(command: CDVInvokedUrlCommand) {
        var pluginResult: CDVPluginResult
        
        guard let arg1 = command.arguments[0] as? String else {
            pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
            commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        let echoedMessage = "\(arg1)"
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: echoedMessage)
        commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc func changeNotification(variableId: String) {
        
        let valor = variablesMap[variableId] ?? ""
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: valor)
        pluginResult?.keepCallback = true
        
        let matchingCallbacks = callbackMap.keys.filter { $0.starts(with: "\(variableId):") }
        matchingCallbacks.forEach { key in
            if let callbackContext = callbackMap[key] {
                print("VALOR: \(valor) TO \(callbackContext)")
                commandDelegate.send(pluginResult, callbackId: callbackContext)
            }
        }
    }
    
    func streamHR() {
        hrDisposable = api.startHrStreaming(deviceId)
            .observe(on: MainScheduler.instance)
            .subscribe(onNext: {
                hrData in
                for sample in hrData {
                    var hr = sample.hr
                    self.variablesMap["heartRate"] = "\(hr)"
                    self.changeNotification(variableId: "heartRate")
                    //NSLog("HEART RATE: \(hr)")
                }
                
            }, onError: { error in
                NSLog("HR stream failed. Reason \(error)")
            }, onCompleted: {
                NSLog("HR stream complete")
            })
    }
    
    func streamECG() {
        ecgDisposable = api.requestStreamSettings(deviceId, feature: PolarDeviceDataType.ecg)
            .asObservable()
            .flatMap { sensorSetting -> Observable<PolarEcgData> in
                return self.api.startEcgStreaming(self.deviceId, settings: sensorSetting.maxSettings())
            }
            .observe(on: MainScheduler.instance)
            .subscribe(onNext: { polarEcgData in
                self.transformPolarEcgData(polarEcgData)
                    .subscribe(
                        onSuccess: { ecg in
                            self.variablesMap["ecg"] = "\(ecg)"
                            self.changeNotification(variableId: "ecg")
                        },
                        onFailure: { error in
                            NSLog("Error \(error)")
                        }
                    )
                    .disposed(by: DisposeBag())
                
            }, onError: { error in
                NSLog("ECG stream failed. Reason \(error)")
            }, onCompleted: {
                NSLog("ECG stream complete")
            })
    }
    
    /*func transformPolarEcgData(_ polarEcgData: PolarEcgData) -> (time: [UInt64], data: [Float]) {
        let timeStamps = polarEcgData.samples.map { $0.timeStamp }
        let waves = polarEcgData.samples.map { Float($0.voltage) / 1000.0 }
        return (time: timeStamps, data: waves)
    }*/
    
    func transformPolarEcgData(_ polarEcgData: PolarEcgData) -> Single<String> {
        return Single.create { single in
            do {
                let timeStamps = polarEcgData.samples.map { $0.timeStamp }
                let voltages = polarEcgData.samples.map { Float($0.voltage) / 1000.0 }
                let jsonDict: [String: Any] = ["time": timeStamps, "data": voltages]
                let jsonData = try JSONSerialization.data(withJSONObject: jsonDict, options: [])
                if let jsonString = String(data: jsonData, encoding: .utf8) {
                    single(.success(jsonString))
                    
                } else {
                    single(.failure(NSError(domain: "Error al convertir datos a cadena", code: 0, userInfo: nil)))
                    
                }
                
            } catch {
                single(.failure(error))
                
            }
            return Disposables.create()
            
        }
        
    }

}

extension PolarBlePlugin : PolarBleApiDeviceInfoObserver {
    func batteryLevelReceived(_ identifier: String, batteryLevel: UInt) {
        //print("battery level updated: \(batteryLevel)")
        variablesMap["batteryLevel"] = batteryLevel.description
        //variablesMap["heartRate"] = "80"
        Task { @MainActor in
            //self.batteryStatusFeature.batteryLevel = batteryLevel
            changeNotification(variableId: "batteryLevel")
            //changeNotification(variableId: "heartRate")
        }
    }

    func disInformationReceived(_ identifier: String, uuid: CBUUID, value: String) {
        print("dis info: \(uuid.uuidString) value: \(value)")
        if(uuid == BleDisClient.SOFTWARE_REVISION_STRING) {
            Task { @MainActor in
                //self.deviceInfoFeature.firmwareVersion = value
            }
        }
    }
}


extension PolarBlePlugin : PolarBleApiDeviceFeaturesObserver {
    func bleSdkFeatureReady(_ identifier: String, feature: PolarBleSdk.PolarBleSdkFeature) {
        NSLog("Feature is ready: \(feature)")
        switch(feature) {
        case .feature_polar_online_streaming:
            Task { @MainActor in
                //self.onlineStreamingFeature.isSupported = true
            }
            api.getAvailableOnlineStreamDataTypes(identifier)
                .observe(on: MainScheduler.instance)
                .subscribe{ e in
                    switch e {
                    case .success(_):
                        self.streamHR()
                        self.streamECG()
                    case .failure(let err):
                        NSLog("Failed to get available online streaming data types: \(err)")
                        //self.somethingFailed(text: "Failed to get available online streaming data types: \(err)")
                    }
                }.disposed(by: disposeBag)
            break
        default:
            NSLog("Feature unused")
        }
    }

    // deprecated
    func hrFeatureReady(_ identifier: String) {
        NSLog("HR ready")
    }

    // deprecated
    func ftpFeatureReady(_ identifier: String) {
        NSLog("FTP ready")
    }

    // deprecated
    func streamingFeaturesReady(_ identifier: String, streamingFeatures: Set<PolarDeviceDataType>) {
        for feature in streamingFeatures {
            NSLog("Feature \(feature) is ready.")
        }
    }

}

